OrbitCore/LinuxEventTracer.cpp

```diff
@@ -0,0 +1,136 @@
+//-----------------------------------
+// Copyright Pierric Gimmig 2013-2019
+//-----------------------------------
+
+
+#include "OrbitProcess.h"
+#include "PrintVar.h"
+#include "LinuxPerfEvent.h"
+#include "LinuxPerfRingBuffer.h"
+#include "LinuxPerfUtils.h"
+#include "LinuxEventTracer.h"
+#include "LinuxEventTracerVisitor.h"
+#include "LinuxPerfEventProcessor.h"
+#include "LinuxUtils.h"
+#include "Capture.h"
+
+#include <vector>
+
+
+//-----------------------------------------------------------------------------
+void LinuxEventTracer::Start()
+{
+    PRINT_FUNC;
+#if __linux__
+    m_ExitRequested = false;
+
+    m_Thread = std::make_shared<std::thread>(&LinuxEventTracer::Run, &m_ExitRequested);
+
+    m_Thread->detach();
+#endif
+}
+
+//-----------------------------------------------------------------------------
+void LinuxEventTracer::Stop()
+{
+    m_ExitRequested = true;
+}
+
+void LinuxEventTracer::Run(bool* a_ExitRequested)
+{
+    // get the number of cpus
+    int cpus = std::stoi(LinuxUtils::ExecuteCommand("nproc"));
+
+    uint64_t tracepointID = LinuxUtils::GetTracePointID("sched", "sched_switch");
+    
+    std::vector<uint32_t> fds;
+
+    // open perf events for all cpus to keep track of process spawning
+    for (int cpu = 0; cpu < cpus; cpu++)
+    {
+        fds.push_back(LinuxPerfUtils::task_event_open(cpu)); 
+        fds.push_back(LinuxPerfUtils::tracepoint_event_open(tracepointID, -1 /*pid*/, cpu)); 
+    }
+
+    std::vector<std::shared_ptr<LinuxPerfRingBuffer>> ring_buffers;
+    for (int32_t fd : fds)
+    {
+        auto ring_buffer = std::make_shared<LinuxPerfRingBuffer>(fd);
+        ring_buffers.push_back(ring_buffer);
+    
+        // start recording via ioctl
+        LinuxPerfUtils::start_capturing(fd);
+    }
+
+    // keep threads in synch.
+    auto threads = LinuxUtils::ListThreads(Capture::GTargetProcess->GetID());
+    for (uint64_t tid : threads) {
+        Capture::GTargetProcess->AddThreadId(tid);
+    }
+
+    LinuxPerfEventProcessor event_buffer(std::make_unique<LinuxEventTracerVisitor>());
+
+    while (!(*a_ExitRequested))
+    {
+        usleep(5000);
```

Comment:
What's the reason for the sleep?

Reply:
Without the sleep, this thread would constantly read the metadata from the buffer in order to check if there are new events. Reading the header would be done in a much higher frequency than new events can even occur, and we would wast a lot of CPU time, basically useless.

Reply:
I would add a comment that explains this in the code.